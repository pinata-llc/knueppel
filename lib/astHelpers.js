"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.idIn = exports.bin = exports.uuidList = exports.buildLogicalExpression = void 0;
function buildLogicalExpression(operator, expressions) {
    var outerLe = {
        type: "LogicalExpression",
        operator: operator,
        left: null,
        right: null,
    };
    var le = outerLe;
    for (var x = 0; x < expressions.length; x++) {
        var exp = expressions[x];
        if (le.left) {
            if (x + 1 < expressions.length) {
                le.right = {
                    type: "LogicalExpression",
                    operator: operator,
                    left: exp,
                    right: null,
                };
                le = le.right;
            }
            else {
                le.right = exp;
            }
        }
        else {
            le.left = exp;
        }
    }
    if (outerLe.left) {
        if (!outerLe.right) {
            return outerLe.left;
        }
        return outerLe;
    }
    return null;
}
exports.buildLogicalExpression = buildLogicalExpression;
var uuidList = function (values) { return ({ type: "UUIDListLiteral", values: values }); };
exports.uuidList = uuidList;
function bin(left, operator, right) {
    return {
        type: "BinaryExpression",
        left: typeof left === "string" ? { type: "Identifier", name: left } : left,
        operator: operator,
        right: typeof right === "object"
            ? right
            : {
                type: "Literal",
                value: right,
            },
    };
}
exports.bin = bin;
function idIn(ids) {
    return bin("id", "in", exports.uuidList(ids));
}
exports.idIn = idIn;
