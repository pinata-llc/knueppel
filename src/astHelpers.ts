import { BinaryOperator } from "./expression/binaryExpression";
import { LogicalOperator } from "./expression/logicalExpression";

export type Expression =
  | BinaryExpression
  | GeoRadiusExpression
  | LogicalExpression
  | Identifier
  | Literal
  | NullLiteral
  | ListLiteral
  | UUIDListLiteral
  | PointLiteral
  | NotExpression
  | AnyExpression;

export interface BinaryExpression {
  type: "BinaryExpression";
  operator: BinaryOperator;
  left: Expression;
  right: Expression;
}

export interface AnyExpression {
  type: "AnyExpression";
  values: ListLiteral | UUIDListLiteral | Identifier;
}

export interface LogicalExpression {
  type: "LogicalExpression";
  operator: LogicalOperator;
  left: Expression;
  right: Expression;
}

export interface GeoRadiusExpression {
  type: "GeoRadiusExpression";
  pointA: Expression;
  radius: Expression;
  unit: "km" | "mi";
  pointB: Expression;
}

export interface Identifier {
  type: "Identifier";
  name: string;
  args?: Expression[];
}

export interface Literal {
  type: "Literal";
  value: string | number | boolean;
}

export interface ListLiteral {
  type: "ListLiteral";
  values: string[] | number[] | boolean[];
}

export interface NullLiteral {
  type: "NullLiteral";
}

export interface UUIDListLiteral {
  type: "UUIDListLiteral";
  values: string[];
}

export interface PointLiteral {
  type: "PointLiteral";
  lat: number;
  long: number;
}

export interface NotExpression {
  type: "NotExpression";
  argument: Expression;
}

interface PartialLE {
  type: "LogicalExpression";
  operator: LogicalOperator;
  left: PartialLE | Expression | null;
  right: PartialLE | Expression | null;
}

export function buildLogicalExpression(
  operator: LogicalOperator,
  expressions: Array<Expression | null>,
): Expression | null {
  const outerLe: PartialLE = {
    type: "LogicalExpression",
    operator,
    left: null,
    right: null,
  };

  let le = outerLe;

  for (let x = 0; x < expressions.length; x++) {
    const exp = expressions[x];

    if (le.left) {
      if (x + 1 < expressions.length) {
        le.right = {
          type: "LogicalExpression",
          operator,
          left: exp,
          right: null,
        };

        le = le.right;
      } else {
        le.right = exp;
      }
    } else {
      le.left = exp;
    }
  }

  if (outerLe.left) {
    if (!outerLe.right) {
      return outerLe.left as LogicalExpression;
    }
    return outerLe as LogicalExpression;
  }

  return null;
}

export const uuidList = (values: string[]): UUIDListLiteral => ({ type: "UUIDListLiteral", values });

export function bin(
  left: Expression | string,
  operator: BinaryOperator,
  right: Expression | string | number,
): BinaryExpression {
  return {
    type: "BinaryExpression",
    left: typeof left === "string" ? { type: "Identifier", name: left } : left,
    operator,
    right:
      typeof right === "object"
        ? right
        : {
            type: "Literal",
            value: right,
          },
  };
}

export function idIn(ids: string[]) {
  return bin("id", "in", uuidList(ids));
}
