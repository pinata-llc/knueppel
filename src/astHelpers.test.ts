import { buildLogicalExpression } from "./astHelpers";

describe("astHelpers", () => {
  describe("buildLogicalExpression", () => {
    it("should create tree", () => {
      const result = buildLogicalExpression("||", [
        {
          type: "BinaryExpression",
          left: { type: "Identifier", name: "productCount" },
          operator: "=",
          right: { type: "Literal", value: 1 },
        },
        {
          type: "BinaryExpression",
          left: { type: "Identifier", name: "productCount" },
          operator: ">",
          right: { type: "Literal", value: 20 },
        },
        {
          type: "BinaryExpression",
          left: { type: "Identifier", name: "startDate" },
          operator: ">",
          right: { type: "Identifier", name: "now" },
        },
      ]);

      expect(result).toMatchSnapshot();
    });
  });
});
