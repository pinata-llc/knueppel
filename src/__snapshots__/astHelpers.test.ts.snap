// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`astHelpers buildLogicalExpression should create tree 1`] = `
Object {
  "left": Object {
    "left": Object {
      "name": "productCount",
      "type": "Identifier",
    },
    "operator": "=",
    "right": Object {
      "type": "Literal",
      "value": 1,
    },
    "type": "BinaryExpression",
  },
  "operator": "||",
  "right": Object {
    "left": Object {
      "left": Object {
        "name": "productCount",
        "type": "Identifier",
      },
      "operator": ">",
      "right": Object {
        "type": "Literal",
        "value": 20,
      },
      "type": "BinaryExpression",
    },
    "operator": "||",
    "right": Object {
      "left": Object {
        "name": "startDate",
        "type": "Identifier",
      },
      "operator": ">",
      "right": Object {
        "name": "now",
        "type": "Identifier",
      },
      "type": "BinaryExpression",
    },
    "type": "LogicalExpression",
  },
  "type": "LogicalExpression",
}
`;
